File: tools/D31-wireless-adb/source/tools/baseline/collect_start_sh_post_evidence.py
"""衔接原交接、后置原件和原verify；仅本地文件与原Java合同。"""
import json
from pathlib import Path
import shlex
import sys
import time

sys.dont_write_bytecode = True
import check_start_sh_handoff as handoff
import start_sh_repair as repair


def preflight(options, java, now):
    value = handoff.check(options['handoff'], options['handoffSha256'], options['deviceId'], java, now, historical=True)
    eb, need = repair.eb, repair.need
    root = Path(options['handoff']).resolve().parent
    prepared = repair.obj(root / 'prepared/prepared-private.json')
    raw = eb.read_bytes(options['handoffResult'])
    need(repair.digest(raw) == options['handoffResultSha256'], 'QUERY_CAPTURE_DIGEST_MISMATCH')
    capture = eb.baseline.strict_json(raw)
    for key in ('handoffSha256', 'preparedSha256', 'planSha256', 'deviceId', 'taskId', 'fixture'):
        need(eb.same(capture[key], value[key]), 'QUERY_CAPTURE_BINDING')
    request, response = capture['queryRequest'], capture['queryResult']
    task = repair.parent(request, response, value['deviceId'], 'root_exec')
    result = task['result']
    need(result.get('action') == 'completed' and type(result.get('exit_code')) is int
         and result['exit_code'] == 0 and result.get('truncated') is False, 'QUERY_OUTPUT_INCOMPLETE')
    apk = '/data/local/d31-remote/releases/' + eb.hash_value(options['apkSha256']) + '/remote.apk'
    argv = shlex.split(request['params']['command'])
    need(len(argv) == 5 and argv[:4] == ['CLASSPATH=' + apk, '/system/bin/app_process', '/system/bin',
         'net.elfradio.d31bootstrap.RemoteRepairCommand'], 'QUERY_COMMAND_SHAPE')
    need(eb.same(eb.baseline.strict_json(argv[4].encode()), value['requests']['query']), 'QUERY_PLAN_BINDING')
    receipt = eb.embedded_json(result['text'])
    need(receipt.get('schema') == 1 and receipt.get('task_id') == value['taskId']
         and receipt.get('plan_sha256') == value['planSha256'], 'RECEIPT_PLAN_BINDING')
    need(receipt.get('verification_scope') == 'FILE_CONTENT_AND_METADATA' and receipt.get('runtime_effect') == 'NOT_CHECKED'
         and receipt.get('system_consistency') == 'NOT_ASSESSED', 'RECEIPT_SCOPE')
    count = eb.integer(receipt['next_event'])
    need(0 < count <= 256 and receipt['state']['phase'] in ('SUCCEEDED', 'REJECTED', 'ROLLED_BACK', 'NEEDS_ATTENTION'), 'TERMINAL_QUERY_REQUIRED')
    journal = '/data/local/d31-remote/repairs/' + value['taskId']
    need(receipt.get('journal') == journal, 'REPAIR_JOURNAL_MISMATCH')
    start, end = repair.stamp(task['started_at']), repair.stamp(task['completed_at'])
    need(end >= start > prepared['preimageProof']['completedAtMs'], 'QUERY_TIME_INVALID')
    need(type(options['version']) is int and options['version'] > 0, 'VERSION_REQUIRED')
    return {'value': value, 'prepared': prepared, 'queryRequest': request, 'queryResult': response,
            'receipt': receipt, 'queryEnd': end, 'journal': journal, 'apk': apk}


def finalize(options, collection, output, java, now):
    context = preflight(options, java, now)
    eb, need = repair.eb, repair.need
    device = context['value']['deviceId']
    root = Path(options['handoff']).resolve().parent
    proof = collection['proofs']
    # 先复核取回父任务与原字节，再组原证据包；不能以本地目标载荷充当后像。
    diagnostic = proof['diagnostic']
    parent_request, parent_result = repair.obj(diagnostic['request']), repair.obj(diagnostic['result'])
    task = repair.parent(parent_request, parent_result, device, 'root_exec')
    result = task['result']
    need(result.get('action') == 'completed' and type(result.get('exit_code')) is int
         and result['exit_code'] == 0 and result.get('truncated') is False, 'DIAGNOSTIC_OUTPUT_INCOMPLETE')
    diagnostic_id, request = eb.command_request(parent_request['params']['command'])
    need(diagnostic_id == collection['diagnosticId'] and eb.same(request, collection['diagnosticRequest']), 'DIAGNOSTIC_BINDING')
    receipt = eb.embedded_json(result['text'])
    report_path = '/data/local/d31-remote/diagnostics/' + diagnostic_id + '/report.json'
    observation, returned = repair.returned(proof['observation'], device, report_path, context['queryEnd'])
    need(returned['startedAtMs'] > repair.stamp(task['completed_at']), 'REPORT_NOT_AFTER_DIAGNOSTIC')
    eb.match_digest(receipt, observation)
    after, after_proof = repair.returned(proof['after_file'], device, repair.TARGET, context['queryEnd'])
    need(after_proof['startedAtMs'] > returned['completedAtMs'], 'AFTER_FILE_ORDER')
    before_record, before_raw, before_objects = repair.bundle(root / 'evidence', context['value']['evidenceSha256'])
    observed = eb.baseline.strict_json(observation)
    report = java.call('compare', {'observation': observed['manifest'], 'firmware': before_objects['firmware'], 'now': now})
    report['inputs'] = {role: {**eb.digest(raw), 'manifestPointer': pointer} for role, raw, pointer in
                        [('observation', observation, '/manifest'), ('firmware', before_raw['firmware'], '')]}
    output.mkdir(mode=0o700, exist_ok=False)
    inputs = output / 'inputs'; inputs.mkdir()
    raw = {'request': eb.encode(request), 'receipt': eb.encode(receipt), 'report': eb.encode(report),
           'observation': observation, 'firmware': before_raw['firmware'], 'firmware_generation': before_raw['firmware_generation'],
           'parent_request': eb.read_bytes(diagnostic['request']), 'parent_result': eb.read_bytes(diagnostic['result'])}
    paths = {}
    for role, data in raw.items():
        paths[role] = str(inputs / (role + '.json')); eb.write_new(Path(paths[role]), data)
    seal = eb.prepare(paths, output / 'after-bundle', fixture=before_record['fixture'], now_ms=now)
    for name in ('queryRequest', 'queryResult'):
        repair.save(output / (name + '.json'), context[name])
    config = {'prepared': str(root / 'prepared/prepared-private.json'), 'prepared_sha256': context['value']['preparedSha256'],
              'before_bundle': str(root / 'evidence'), 'target_payload': str(root / 'prepared/target-payload.sh'),
              'query_request': str(output / 'queryRequest.json'), 'query_result': str(output / 'queryResult.json'),
              'device_plan': proof['plan'], 'events': [proof[f'event-{i:06d}'] for i in range(context['receipt']['next_event'])],
              'after_bundle': str(output / 'after-bundle'), 'after_evidence_sha256': seal['sha256'],
              'after_file': proof['after_file'], 'boot': proof['boot']}
    repair.save(output / 'verify-config-private.json', config)
    verified = repair.verify(config, output / 'verified', java, now)
    return {'closedLoopVerified': verified['closed'], 'fixture': verified['fixture'], 'status': verified['status'],
            'runtimeVerification': verified['runtimeVerification'], 'systemConsistency': verified['systemConsistency'],
            'verifyConfig': str(output / 'verify-config-private.json')}


if __name__ == '__main__':
    try:
        data = json.load(sys.stdin)
        options = data['options']
        java = repair.Java(options['java'], options['classpath'])
        now = time.time_ns() // 1000000
        if data['action'] == 'preflight':
            result = preflight(options, java, now)
        else:
            repair.need(data['action'] == 'finalize', 'ACTION_INVALID')
            result = finalize(options, data['collection'], Path(data['output']).resolve(), java, now)
        print(json.dumps(result))
    except Exception as error:
        code = str(error)
        print(code if code and all(c in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_' for c in code)
              else 'LOCAL_POST_EVIDENCE_REJECTED', file=sys.stderr)
        raise SystemExit(2) from None

File: tools/D31-wireless-adb/source/tools/Collect-StartShPostEvidence.mjs
import fs from 'node:fs';
import path from 'node:path';
import {fileURLToPath} from 'node:url';
import {spawnSync} from 'node:child_process';
import {randomUUID, randomBytes} from 'node:crypto';
import {isDeepStrictEqual as equal} from 'node:util';
import {QueueStore, QueueError, hash, requireThat as need} from './fault-transfer/QueueStore.mjs';
import {WebTransport} from './fault-transfer/WebTransport.mjs';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const TARGET = '/data/local/d31-system-support/start.sh';
const BOOT = '/system/bin/cat /proc/sys/kernel/random/boot_id';
const ACTIVE = '/system/bin/cat /data/local/d31-remote/runtime/active.json';
const terminal = new Set(['success', 'failed', 'rejected', 'expired', 'cancelled']);
const read = p => JSON.parse(fs.readFileSync(p, 'utf8').replace(/^\uFEFF/, ''));
const quote = s => "'" + s.replaceAll("'", "'\\''") + "'";
const stamp = value => {
  const time = typeof value === 'string' && value.endsWith('Z') ? Date.parse(value) : value;
  need(Number.isSafeInteger(time) && time >= 0, 'TASK_TIME_INVALID'); return time;
};

function local(data) {
  const p = spawnSync(data.options.python || 'python', ['-B', path.join(HERE, 'baseline/collect_start_sh_post_evidence.py')],
    {input: JSON.stringify(data), encoding: 'utf8', windowsHide: true, timeout: 180000, maxBuffer: 1048576});
  need(p.status === 0, /^[A-Z0-9_]+$/.test(p.stderr?.trim() || '') ? p.stderr.trim() : 'LOCAL_VERIFY_FAILED');
  return JSON.parse(p.stdout);
}

function parent(task, request, after) {
  need(task?.id === request.id && task.type === request.type, 'PARENT_ID_TYPE_MISMATCH');
  need(task.device_id === undefined || task.device_id === request.device_id, 'PARENT_DEVICE_MISMATCH');
  need(task.params === undefined || equal(task.params, request.params), 'PARENT_PARAMS_MISMATCH');
  need(typeof task.state === 'string', 'PARENT_STATE_MISSING');
  if (!terminal.has(task.state)) return false;
  need(task.state === 'success', 'PARENT_FAILED');
  need(stamp(task.started_at) > after && stamp(task.completed_at) >= stamp(task.started_at), 'POST_TASK_TIME_INVALID');
  const r = task.result;
  if (request.type === 'root_exec') {
    need(r?.action === 'completed' && r.exit_code === 0 && r.truncated === false && typeof r.text === 'string', 'ROOT_OUTPUT_INCOMPLETE');
  } else need(r?.action === 'uploaded' && /^[a-f0-9]{64}$/.test(r.sha256 || '')
    && Number.isSafeInteger(r.bytes) && r.bytes > 0 && r.bytes <= 8388608, 'FILE_RETURN_INVALID');
  return true;
}

// 独立延续已保存的query；没有submit/run、上传或修改目标的入口。
export async function collect(options, {transport, transportFactory, simulation = false, now = Date.now,
  wait = ms => new Promise(resolve => setTimeout(resolve, ms))} = {}) {
  const context = local({action: 'preflight', options});
  if (!options.execute) return {status: 'LOCAL_PREVIEW', eventCount: context.receipt.next_event, closedLoopVerified: false};
  need(!context.value.fixture || simulation, 'SYNTHETIC_EVIDENCE_CANNOT_EXECUTE');
  need(options.state, 'PERSISTENT_STATE_DIRECTORY_REQUIRED');
  const store = new QueueStore(options.state, {python: options.python || 'python'});
  return store.withLock(async () => {
    const binding = {handoff: options.handoffSha256, query: options.handoffResultSha256,
      apk: options.apkSha256, version: options.version, device: options.deviceId};
    let state = store.load();
    if (state) need(equal(state.binding, binding), 'STATE_BINDING_MISMATCH');
    else {
      const diagnosticId = randomBytes(32).toString('hex');
      const diagnosticRequest = {operation: 'manifest', scope: TARGET,
        identity: {...context.prepared.observationIdentity, snapshotId: 'readonly-' + randomUUID().replaceAll('-', '')}};
      state = {binding, diagnosticId, diagnosticRequest, steps: {}}; store.save(state);
    }
    const save = () => store.save(state);
    const io = () => ({signal: AbortSignal.timeout(25000)});
    const t = transport || transportFactory();
    async function task(label, type, params, after) {
      let step = state.steps[label];
      if (!step) {
        const request = {device_id: options.deviceId, id: 'd31-post-' + randomUUID(), type, params, expires_at: now() + 600000};
        step = state.steps[label] = {request, intent: false,
          requestFile: store.artifact(label + '-request-private', request)}; save();
      }
      need(equal(read(store.file(step.requestFile)), step.request) && step.request.type === type
        && step.request.device_id === options.deviceId && equal(step.request.params, params), 'PERSISTED_REQUEST_CHANGED');
      let current;
      if (step.resultFile) current = read(store.file(step.resultFile)).task;
      else {
        if (step.intent) current = await t.queryTask(options.deviceId, step.request.id, io());
        else {
          need(now() < step.request.expires_at, 'REQUEST_EXPIRED');
          step.intent = true; save();
          try { current = await t.enqueue(step.request, io()); }
          catch { current = await t.queryTask(options.deviceId, step.request.id, io()); }
        }
        // 权威明确未入队才补投同一请求；未知下载或未知查询不产生新编号。
        if (current === null) {
          need(now() < step.request.expires_at, 'UNKNOWN_PARENT_EXPIRED');
          try { current = await t.enqueue(step.request, io()); }
          catch { current = await t.queryTask(options.deviceId, step.request.id, io()); }
        }
        for (let i = 0; i < 60; i++) {
          need(current !== null, 'UNKNOWN_PARENT_RESUME_ORIGINAL_ID');
          const file = store.artifact(label + '-response-private', {ok: true, task: current});
          if (terminal.has(current.state)) { step.resultFile = file; save(); }
          if (parent(current, step.request, after)) break;
          need(i < 59, 'PARENT_PENDING_RESUME_ORIGINAL_ID');
          await wait(2000); current = await t.queryTask(options.deviceId, step.request.id, io());
        }
      }
      need(parent(current, step.request, after), 'PARENT_PENDING_RESUME_ORIGINAL_ID');
      return {step, result: current.result, ended: stamp(current.completed_at)};
    }
    const root = (label, command, after, timeout = 30) => task(label, 'root_exec', {command, cwd: '/', timeout}, after);
    async function deliver(label, step, r) {
      const binding = {deviceId: options.deviceId, taskId: step.request.id, path: step.request.params.path,
        size: r.bytes, sha256: r.sha256, file: step.file,
        requestSha256: hash(fs.readFileSync(store.file(step.requestFile))),
        resultSha256: hash(fs.readFileSync(store.file(step.resultFile)))};
      if (!step.delivery) {
        step.delivery = {binding, proof: store.artifact(label + '-local-verified-private', binding),
          state: 'PENDING', attempts: 0, nextAt: 0}; save();
      }
      const d = step.delivery;
      const verifyLocal = () => {
        need(equal(d.binding, binding) && equal(read(store.file(d.proof)), binding)
          && equal(read(store.file(step.requestFile)), step.request)
          && hash(fs.readFileSync(store.file(step.resultFile))) === binding.resultSha256
          && store.bundleMatches(step.file, r), 'DELIVERY_LOCAL_PROOF_MISMATCH');
      };
      verifyLocal();
      if (d.state === 'DELIVERED' || d.nextAt > now()) return;
      // 存储接收回执不代表修复验收；原字节及父任务证据先持久化并再回读。
      d.attempts++; d.nextAt = now() + 2000; save(); verifyLocal();
      let response;
      try {
        response = await t.received(binding.deviceId, binding.taskId, {size: binding.size, sha256: binding.sha256}, io());
        store.artifact(label + '-received-private', response);
        need(response?.ok === true && ((response.purged === true && response.cleanup_pending !== true)
          || (response.cleanup_pending === true && response.purged !== true)), 'RETURN_RECEIPT_RESPONSE_INVALID');
      } catch (error) {
        if (!(error instanceof QueueError)) throw error;
        d.lastError = {code: error.code, at: now()};
        d.nextAt = Math.max(now() + Math.min(60000, 2000 * 2 ** Math.min(5, d.attempts)),
          Number.isSafeInteger(error.retryAt) ? error.retryAt : 0); save(); return;
      }
      d.state = 'DELIVERED'; d.purged = response.purged === true;
      d.cleanupPending = response.cleanup_pending === true; d.nextAt = 0; delete d.lastError; save();
    }
    async function get(label, remote, after, limit) {
      const item = await task(label, 'get_file', {path: remote, allow_cellular: false}, after);
      const {step, result: r} = item;
      need(r.bytes <= limit, 'RETURN_SIZE_LIMIT');
      if (!step.file) {
        const meta = await t.metadata(options.deviceId, step.request.id, io());
        store.artifact(label + '-metadata-private', meta);
        need(meta.ok === true && meta.file?.device_id === options.deviceId && meta.file.task_id === step.request.id
          && meta.file.size === r.bytes && meta.file.sha256 === r.sha256, 'METADATA_MISMATCH');
        store.capacity(r.bytes);
        const temporary = label + '-' + randomUUID() + '.part';
        await t.download(options.deviceId, step.request.id, {filename: store.file(temporary), bytes: r.bytes,
          signal: AbortSignal.timeout(120000), onBytes: () => {}});
        need(store.bundleMatches(temporary, r), 'DOWNLOAD_DIGEST_MISMATCH');
        step.file = temporary; save();
      }
      need(store.bundleMatches(step.file, r), 'SAVED_DOWNLOAD_CHANGED');
      await deliver(label, step, r);
      return {...item, bytes: fs.readFileSync(store.file(step.file))};
    }
    function active(item) {
      const a = JSON.parse(item.result.text);
      need(a.versionCode === options.version && a.sha256 === options.apkSha256 && a.path === context.apk, 'ACTIVE_APK_MISMATCH');
    }
    function boot(item) { need(item.result.text.trim() === context.value.bootId, 'BOOT_CHANGED'); }
    active(await root('active-before', ACTIVE, context.queryEnd));
    boot(await root('boot-before', BOOT, context.queryEnd));
    const plan = await get('plan', context.journal + '/plan.json', context.queryEnd, 131072);
    let head = hash(plan.bytes), last;
    for (let i = 0; i < context.receipt.next_event; i++) {
      const event = await get('event-' + String(i).padStart(6, '0'), context.journal + '/' + String(i).padStart(6, '0') + '.json', context.queryEnd, 131072);
      const obj = JSON.parse(event.bytes);
      need(obj.sequence === i && obj.previous_sha256 === head && obj.plan_sha256 === context.value.planSha256, 'EVENT_CHAIN_MISMATCH');
      head = hash(event.bytes); last = obj.state;
    }
    need(equal(last, context.receipt.state), 'QUERY_EVENT_PREFIX_MISMATCH');
    const reportPath = '/data/local/d31-remote/diagnostics/' + state.diagnosticId + '/report.json';
    const diagnostic = await root('diagnostic', 'CLASSPATH=' + quote(context.apk)
      + ' /system/bin/app_process /system/bin net.elfradio.d31bootstrap.RemoteDiagnosticCommand '
      + quote(state.diagnosticId) + ' ' + quote(JSON.stringify(state.diagnosticRequest)), context.queryEnd, 120);
    const receipt = JSON.parse(diagnostic.result.text);
    need(receipt.id === state.diagnosticId && receipt.path === reportPath && receipt.state === 'completed', 'DIAGNOSTIC_RECEIPT_INVALID');
    const observation = await get('observation', reportPath, diagnostic.ended, 8388608);
    need(receipt.sha256 === hash(observation.bytes) && receipt.bytes === observation.bytes.length, 'REPORT_DIGEST_MISMATCH');
    const m = JSON.parse(observation.bytes).manifest;
    for (const [key, value] of Object.entries(state.diagnosticRequest.identity)) need(equal(m[key], value), 'OBSERVATION_IDENTITY_CHANGED');
    need(m.snapshotId !== context.prepared.before.snapshotId && state.diagnosticId !== context.prepared.before.diagnosticId, 'CACHED_DIAGNOSTIC_REUSED');
    const after = await get('after_file', TARGET, observation.ended, 4096);
    boot(await root('boot', BOOT, after.ended));
    active(await root('active-after', ACTIVE, after.ended));
    const proofs = Object.fromEntries(Object.entries(state.steps).map(([label, step]) => [label, {
      request: store.file(step.requestFile), result: store.file(step.resultFile), ...(step.file ? {file: store.file(step.file)} : {}),
    }]));
    const collection = {diagnosticId: state.diagnosticId, diagnosticRequest: state.diagnosticRequest, proofs};
    store.artifact('collection-private', collection);
    // 每次离线复核使用新目录；异常中断不覆盖先前半成品，也不重跑修复或诊断。
    const output = path.join(path.dirname(store.directory), path.basename(store.directory) + '-verify-' + randomUUID());
    const verified = local({action: 'finalize', options, collection, output});
    const deliveries = Object.values(state.steps).filter(step => step.delivery).map(step => step.delivery);
    const returnCleanup = {delivered: deliveries.filter(d => d.state === 'DELIVERED').length,
      pending: deliveries.filter(d => d.state !== 'DELIVERED').length,
      cleanupPending: deliveries.filter(d => d.cleanupPending).length};
    const result = {...verified, simulation, executed: !simulation, returnCleanup};
    store.artifact('post-evidence-result-private', result); return result;
  });
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  try {
    const [config, flag] = process.argv.slice(2);
    need(config && (flag === undefined || flag === '--execute') && process.argv.length <= 4, 'CLI_ARGUMENT_INVALID');
    const options = {...read(config), execute: flag === '--execute'};
    const result = await collect(options, {transportFactory: () => {
      need(options.session, 'SESSION_REQUIRED'); return new WebTransport({session: read(options.session)});
    }});
    console.log(JSON.stringify({status: result.status, closedLoopVerified: result.closedLoopVerified,
      fixture: result.fixture, executed: result.executed || false, returnCleanup: result.returnCleanup}));
  } catch (error) {
    console.error('后置证据停止，保留原号和原件：' + (/^[A-Z0-9_]+$/.test(error.message) ? error.message : 'POST_EVIDENCE_FAILED'));
    process.exitCode = 2;
  }
}
